enhancements/multi-arch/multiarch-manager-operator.md

```diff
@@ -0,0 +1,845 @@
+
+The multiarch-manager-operator will leverage the scheduling readiness feature
+and the mutable pod scheduling directives to update the node affinity of pods
+with the set of architectures they support. When a pod is submitted to the API
+server, a mutating admission webhook intercepts the pod and adds a scheduling
+gate to the pod spec.
+
+The `PodPlacementController` watches the pod resource. Suppose any pod has the
+scheduling gate set. In that case, the images in the pod spec are inspected, a
+subset of architectures that satisfy all containers in the pod is built, and the
+pod spec is patched with the additional node affinity constraints. If any
+existing node selector requirements are set with key `kubernetes.io/arch`,
+the `PodPlacementController` does not override the behavior previously
+pushed by other stakeholders, leaving it as complies with KEP-3838[^2].
+
+The multiarch-manager-operator will be available as part of the OperatorHub, as
+a secondary operator managed by the Operator Lifecycle Manager (OLM). It will be
+a cluster-scoped operator that only cluster admins can install and configure.
+
+Its development is expected to ship in the following phases:
+
+- Phase 1 - Technology Preview (TP):
+  - The pod placement controller will be supported in High Availability (HA)
+    mode
+  - The webhook will be a stateless component
+  - The pod placement controller will support namespace filtering to skip pods
+    that should not be mutated
+  - No support for customized network scenarios (restricted, proxied,
+    disconnected...)
+  - No caching facilities for the image metadata
+  - No graceful uninstallation support: any remaining gated pods have to be
+    manually patched to resume scheduling after uninstallation
+- Phase 2 - General Availability (GA):
+  - support for customized network scenarios (restricted, proxied,
+    disconnected...)
+  - support mirror registries through ICSPs/IDMSs
+  - implement caching facilities for the image metadata
+- Phase 3:
+  - Future performance optimizations
+
+### Workflow Description
+The following section will highlight the main workflows the operator will be
+responsible for from a high-level user perspective.
+
+#### Operator installation
+1. The cluster-admin submits a Subscription object to install the
+   multiarch-manager-operator, either via CLI or the OperatorHub
+2. The cluster admin submits the default `PodPlacementConfig` Custom Resource
+   (CR) adding their specific configuration related to initializing the
+   multiarch-manager-operator's components; the cluster admin can customize the
+   CR by using a namespace selector to control pod filtering and logging levels
+4. The operator reconciles the CR. Its information supports the
+   `PodPlacementController` controller and the `SchedulingGateMutatingWebhook`
+   mutating admission webhook
+
+#### Pod Admission: gating
+1. The user submits a pod
+2. The mutating admission webhook is queried and decides whether to add the
```

Comment:
Hi @ingvagabund, as far as we know, there shouldn't be an issue for other schedulers unless they are based on older Kubernetes APIs. 

In such cases of schedulers built on top of older k8s APIs than the running ones (>=1.27), the scheduler will not know about the schedulingReadiness feature. Still, it will not schedule the pod in any case as the DefaultBinder plugin will get an error from the kube-apiserver, and fail until the schedulingGates are removed from our controller.

Moreover, shouldn't we recommend the users not use a custom-scheduler built on top of API versions that are non-compatible with the Kubernetes in their cluster?

Are these the cases you are concerned about?

Reply:
I was stressing the case where two different schedulers operate over the same namespace. Where a user is interested in mutating only pods scheduled by one of the schedulers, ignoring pods of the other one.

One can always generate pods that do not have two or more different values for .spec.schedulerName inside the same namespaces. So, this is not something critical to deal with now. Just another use case that might be interesting for users.

Reply:
It seems a good point to me, too @ingvagabund.

I wonder if we can use a similar approach to the namespace selectors by providing a pod selector.
We wouldn't bind to the schedulerName, but provide a second-level, more fine-grained approach to include or exclude pods based on their labels.

I'd add this as an open question.